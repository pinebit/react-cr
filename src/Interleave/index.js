import React from 'react';
import PropTypes from 'prop-types';
import Wrapper from '../Wrapper';

const Interleave = ({ separator, first, last, children, ...wrapperProps }) => {
  if (!children || !separator) {
    return null;
  }

  const content = [];
  children.forEach((child, index) => {
    const key = child.key;
    if (key) {
      content.push(child);
      const lastChild = index === children.length - 1;
      if (!lastChild) {
        content.push(React.cloneElement(separator, { key: `${key}_separator` }));
      }
    }
  });

  return (
    <Wrapper {...wrapperProps}>
      {first && React.cloneElement(separator, { key: 'first_separator' })}
      {content}
      {last && React.cloneElement(separator, { key: 'last_separator' })}
    </Wrapper>
  );
};

Interleave.propTypes = {
  separator: PropTypes.node.isRequired,
  first: PropTypes.bool,
  last: PropTypes.bool,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired
};

export default Interleave;
