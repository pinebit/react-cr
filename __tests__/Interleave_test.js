import React from 'react';
import renderer from 'react-test-renderer';
import Interleave from '../src/Interleave';

it('Interleave renders children with separator', () => {
  const tree = renderer.create(
    <Interleave separator={<span>,</span>}>
      <span key="a">A</span>
      <span key="b">B</span>
      <span key="c">C</span>
    </Interleave>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('Interleave renders children with separator and prepend the separator', () => {
  const tree = renderer.create(
    <Interleave separator={<span>,</span>} first>
      <span key="a">A</span>
      <span key="b">B</span>
      <span key="c">C</span>
    </Interleave>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('Interleave renders children with separator and append the separator', () => {
  const tree = renderer.create(
    <Interleave separator={<span>,</span>} last>
      <span key="a">A</span>
      <span key="b">B</span>
      <span key="c">C</span>
    </Interleave>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('Interleave does not render children without keys', () => {
  const tree = renderer.create(
    <Interleave separator={<span>,</span>}>
      <span>A</span>
      <span>B</span>
      <span>C</span>
    </Interleave>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('Interleave does not render with empty separator', () => {
  /* eslint-disable */
  console.error = function () {};
  const tree = renderer.create(
    <Interleave>
      <span key="abc">ABC</span>
    </Interleave>
  ).toJSON();
  /* eslint-enable */

  expect(tree).toMatchSnapshot();
});

it('Interleave does not render without children', () => {
  /* eslint-disable */
  console.error = function () {};
  const tree = renderer.create(
    <Interleave separator={<span>*</span>}>
    </Interleave>
  ).toJSON();
  /* eslint-enable */

  expect(tree).toMatchSnapshot();
});
