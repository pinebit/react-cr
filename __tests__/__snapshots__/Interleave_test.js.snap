// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Interleave does not render children without keys 1`] = `<div />`;

exports[`Interleave does not render with empty separator 1`] = `null`;

exports[`Interleave does not render without children 1`] = `null`;

exports[`Interleave renders children with separator 1`] = `
<div>
  <span>
    A
  </span>
  <span>
    ,
  </span>
  <span>
    B
  </span>
  <span>
    ,
  </span>
  <span>
    C
  </span>
</div>
`;

exports[`Interleave renders children with separator and append the separator 1`] = `
<div>
  <span>
    A
  </span>
  <span>
    ,
  </span>
  <span>
    B
  </span>
  <span>
    ,
  </span>
  <span>
    C
  </span>
  <span>
    ,
  </span>
</div>
`;

exports[`Interleave renders children with separator and prepend the separator 1`] = `
<div>
  <span>
    ,
  </span>
  <span>
    A
  </span>
  <span>
    ,
  </span>
  <span>
    B
  </span>
  <span>
    ,
  </span>
  <span>
    C
  </span>
</div>
`;
